### PR TITLE
Feature: Add Include/Exclude Class regular expression matcher

### PR DESCRIPTION
**Pull request by ying-jeanne**
### What's being changed:
/claim #2481

Never used bounty before, just for fun, thank you for building this nice project.

This PR aims to address [issue #2481](https://github.com/weaviate/weaviate/issues/2481).

I think it is just a bad idea trying to guess whether users intend to use an exact match or a regular expression. So in order to distinguish the intention and give more flexibility on settings, whether we could change `include` list to a list of struct:
```json
{
  "include": [
      {
          "expression": "foo",
          "is_regex": false
      }
  ]
}
```
This approach offers clarity in distinguishing between exact matches and regular expressions. However, it may increase verbosity.

or adding a new array:
```json
{
    "include_patterns": []{}
}
```
This alternative maintains backward compatibility with existing APIs and reduces verbosity. This PR use the second approach, if the direction seems ok, i would update docs and examples accordingly, please let me know.

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [x] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	@if [ ! -f $(GOFUMPT) ]; then \
+		echo "Installing gofumpt..."; \
+		go install mvdan.cc/gofumpt@latest; \
+	else \
+		echo "gofumpt is already installed."; \
+	fi
+
+# Target to format code using gofumpt
+.PHONY: install-gofumpt format
+format: install-gofumpt
```
Comment (ying-jeanne) [inline Makefile:15]: add make format and make generate-code, I used them to fix CI.

**Comment by ying-jeanne** [comment]
After running make generate-code, sonar failed for duplication, i think we should exclude certain folder from duplication check. I am not very familiar with sonar code analysis config.

**Comment by sebawita** [comment]
Hi @ying-jeanne, thank you for your contribution.

I like the first approach,
But can we make it backwards compatible?

If yes, we should be able to cover the following scenarios:

## Fixed names (backwards compatible)

Provide a list of specific collection names (not regex) to include/exclude

```
{
  "include": [ "cars_fast", "cars_safe", "bikes_fast", "bikes_safe", "helicopters"]
}
```

## Single/Multiple Regex
```
{
  "include": [
      {
          "expression": "cars*",
          "is_regex": true
      },
  ]
}
```

```
{
  "include": [
      {
          "expression": "cars*",
          "is_regex": true
      },
      {
          "expression": "bikes*",
          "is_regex": true
      },
  ]
}
```

## Mixed - regex and fixed names
```
{
  "include": [
      {
          "expression": "cars*",
          "is_regex": true
      },
      "helicopters",
  ]
}
```

**Comment by sebawita** [comment]
Hi @ying-jeanne, thank you for your contribution.

I went on a round trip thinking about the best approach. Usually, I like our APIs to be explicit – @dirkkul can vouch for that 😉 

However, in this case, the list of characters that you can use for a collection name is limited to `[a-Z0-9_]`, so we could quite safely treat all inputs as a regex. And if it is an exact collection name, that would still work.

We can cover all sorts of scenarios with the following syntax:

## Fixed names (backwards compatible)

Provide a list of specific collection names (not regex) to include/exclude

```
{
  "include": [ "cars_fast", "cars_safe", "bikes_fast", "bikes_safe", "helicopters"]
}
```

## Single/Multiple Regex
```
{
  "include": ["cars*"]
}
```

```
{
  "include": [
    "cars*",
    "bikes*"
  ]
}
```

## Mixed - regex and fixed names
```
{
  "include": [
      "cars*",
       "helicopters"
  ]
}
```

**Comment by dirkkul** [comment]
Closing for lack of response

**Pull request closed without merging**